tidy3d/components/source/freq_range.py

```diff
@@ -0,0 +1,217 @@
+        
+        # extract frequency-related info
+        freq0  = 0.5 * (fmax + fmin)   # extract central freq
+        fwidth = 0.5 * (fmax - fmin)   # extract bandwidth
+        return cls(freq0=freq0, fwidth=fwidth)
+
+    @classmethod
+    def from_wavelength(cls, wvl0: float, wvl_width: float) -> FreqRange:
+        """
+            method ``from_wavelength()`` updated instance of class ``FreqRange`` by reassigning new 
+            frequency- and wavelength-related parameters;
+
+            NB: central frequency never corresponds to central wavelength!
+            ``lda0 = (lda_min + lda_max) / 2`` implies that ``freq0 != (fmin + fmax) / 2`` and vise versa
+
+            Parameters
+            ----------
+            wvl0 : float
+                Real-valued central wavelength  
+            wvl_width : float
+                Real-valued wavelength range 
+
+            Returns
+            -------
+            FreqRange
+                An instance of ``FreqRange`` defined by central wavelength ``wvl0`` and wavelength range ``wvl_width``
+        """
+
+        # calculate lowest and highest frequencies
+        fmin = td_const.C_0 / (wvl0 + wvl_width)
+        fmax = td_const.C_0 / (wvl0 - wvl_width)
+
+        return cls.from_freq_interval(fmin=fmin, fmax=fmax)
+
+    @classmethod
+    def from_wvl_interval(cls, wvl_min: float, wvl_max: float) -> FreqRange:
+        """
+            method ``from_wvl_interval()`` updated instance of class ``FreqRange`` by reassigning new 
+            frequency- and wavelength-related parameters; 
+
+            NB: central frequency never corresponds to central wavelength!
+            ``lda0 = (lda_min + lda_max) / 2`` implies that ``freq0 != (fmin + fmax) / 2``
+
+            Parameters
+            ----------
+            wvl_min : float
+                The lowest wavelength of interest. 
+            wvl_max : float
+                The longest wavelength of interest.
+
+            Returns
+            -------
+            FreqRange
+                An instance of ``FreqRange`` defined by the wavelength interval [``wvl_min``, ``wvl_max``]
+        """
+
+        # convert wavelength intervals to frequency interval
+        fmax = td_const.C_0 / wvl_min
+        fmin = td_const.C_0 / wvl_max
+
+        return cls.from_freq_interval(fmin=fmin, fmax=fmax)
+    
+
+    def freqs(self, num_points: int)->NDArray[np.float64]:
+        """
+            method ``freqs()`` returns a numpy array of ``num_point`` frequencies uniformly
+            sampled from the specified frequency range; 
+            if ``num_points == 1`` method returns the central frequency ``freq0``
+
+            Parameters
+            ----------
+            num_points : int
+                Number of frequency points in a frequency range of interest.
+
+            Returns
+            -------
+            np.ndarray
+                a numpy array of uniformly distributed frequency samples in a frequency range of interest
```

Comment:
please do one last readtthough and make sure that there are periods at the end of the field and argument description sentences like this one.

(I know it's super tedious :) , soon we should have an AI bot that will actually catch these for us!)